### DISCUSSION TITLE
Can dosemu send a SysReq keypress condition to DOS?

### DISCUSSION THREAD

**Discussion by ecm-pushbx** [dosemu2/dosemu2#1761] [General]
I'm working on porting over https://hg.pushbx.org/ecm/fddebug/rev/e996e6ff7fb1#l1.216 to lDebug. However, both on dosemu2 and on a hardware machine I am unable to trigger either [the interrupt 15h service 8501h call](https://fd.lod.bz/rbil/interrup/bios/1585.html#1558) or the interrupt 09h (IRQ #1) call that leaves [the keyboard flags](https://fd.lod.bz/rbil/memory/keyboard/m00400018.html) at `byte [40h:18h]` with bit 2 (mask 4) set.

On dosemu2 one problem is that the window never appears to receive the SysReq key as it is intercepted by the window manager (or something like that) and creates a screenshot. However, disabling this in "Keyboard Shortcuts" still doesn't allow dosemu2 to receive it, or it might be dosemu ignores it.

Both methods also seem to fail on my hardware machine as I mentioned. Suggestions are welcome.

**Comment by ecm-pushbx** [comment]
Also tried it in 86Box and this also seems to fail.

**Comment by stsp** [comment]
As far as I can recall, sysrq sends
press and release immediately, so
there was no way to detect when it
is pressed.
Possibly this can work if you switch
kbd to the XT protocol?

**Comment by Baron-von-Riedesel** [comment]
> As far as I can recall, sysrq sends press and release immediately,
> so there was no way to detect when it is pressed.

That's not quite correct. To press and release SysRq sends 2 codes, 0x54 and 0xD4 - and since both codes generate an interrupt, it should be possible to detect sysreq between those 2 interrupts ( it's true, however, that 0x54 is often sent a bit "lately", that is, when the key is released - but it's always sent BEFORE 0xD4 ).

Keyboard emulation is generally a pretty weak point in all emulators - DosEmu, DosBox, DosBox-X, PCem, V86Box, ...

As far as DosEmu2 is concerned: is it supposed to emulate the keyboard interface at the "port" level ( ports 0x60/0x64) or just at the BIOS level? 

Im asking because I have an old, self-written German keyboard driver; it works in Qemu, but hangs if loaded in DosEmu2.

**Reply by stsp** [reply]
> That's not quite correct.

Someone needs to run norton ndiags
on a bare-metal machine, run keyboard
test and see how SysRq behaves.
I recall that it was sending press and
release at once, but I may be wrong.
@jschwartzenberg has the bare-metal
system, so maybe he can test? :)

> As far as DosEmu2 is concerned: is it supposed to emulate the keyboard interface at the "port" level

Of course it does.

**Reply by jschwartzenberg** [reply]
Could you link a copy of that?

**Reply by stsp** [reply]
Its in Norton Utilities.
But I think the case already
settled, thanks for a head up! :)

**Comment by Baron-von-Riedesel** [comment]
> I recall that it was sending press and release at once, but I may be wrong.

No, that's more or less correct. DosEmu sends both 0x54 and 0xD4

I have attached a trivial program that intercepts int 15h, ah=4Fh and displays the codes received there.

So the only reason that's remaining why the FD debugxv doesn't detect SysReq  is because DosEmu forgets to set bit 2 in 0040h:0018h when receiving code 0x54 and resetting that bit when receiving 0xD4.

[GETINT15.EXE.zip](https://github.com/dosemu2/dosemu2/files/9499456/GETINT15.EXE.zip)

**Reply by stsp** [reply]
So do I understand correctly
that the only place where you
can read 0040h:0018h bit 2
is an int15 handler?

**Reply by Baron-von-Riedesel** [reply]
> So do I understand correctly that the only place where you can read 0040h:0018h bit 2 is an int15 handler?

Not precisely - FD debugxv hooks int 09:
```
int09 proc
    pushf
    call cs:[oldint09]
;--- here check 0040:0018, bit 2
    iret
```

That's better, IMO, because one doesn't have to step thru keyboard driver code and one can be sure that EOI has been sent to the PIC.

**Comment by Baron-von-Riedesel** [comment]
Here's a little DOS TSR that cures DosEmu's missing SysReq support

[demusreq.zip](https://github.com/dosemu2/dosemu2/files/9503343/demusreq.zip)

**Reply by stsp** [reply]
You would spend fewer time
preparing a patch to dosemu2
instead, wouldn't you? :)

**Comment by Baron-von-Riedesel** [comment]
> You would spend fewer time preparing a patch to dosemu2 instead, wouldn't you? :)

Perhaps :)). OTOH, to cure it on the DOS side may work for other emulators as well. Although, I tried that getint15 utility in those "other" emulators, and the results I got for pressing/releasing SysReq are - ludicrous.

**Comment by ecm-pushbx** [comment]
From the other thread at https://github.com/dosemu2/dosemu2/issues/1780

> I did test now with Alt+Print/SysReq. The key is labelled "Druck" (Print) on top and "S-Abf" (S-Req) on the front.
> 
> Using an lDebug build with the build options as in `./mak.sh -D_CATCHSYSREQ -D_SYSREQINT=15h` I can get the expected int 15.8501 call, both prior to this commit and after it. However, apparently the int 9 handler is not done with its I/O at this point because the keyboard appears to "lock up" and not accept any more input. I worked around this by running the command `rc.replace g; tp 100` and then `rc`, then interrupting the running debuggee using Alt+SysReq. This suffices to finish dosemu's int 9 handler and regain control in the debugger (keyboard input usable). The test program I used is at https://pushbx.org/ecm/test/20220907/testwait.com (No source, entered into the debugger and saved as an executable.) It loops for about 8 seconds when running.
> 
> An lDebug build with `-D_CATCHSYSREQ -D_SYSREQINT=09h` did not work before, but does work with this commit. Unlike the int 15 method this one appears to finish all keyboard I/O in the called original int 9 handler. So that's definitely more comfortable.
> 
> I also tried Alt+SysReq on my physical machine now (AwardBIOS), and both methods also work for this. However, unlike dosemu2, it appears that the ROM-BIOS's int 9 handler finishes sending EOI and whatever other I/O it does before it calls down to int 15.8501. So both methods are readily usable on this machine, unlike dosemu2 for now.

**Comment by stsp** [comment]
But this was fixed, isn't it?

**Reply by ecm-pushbx** [reply]
For my purposes yes.

**Comment by stsp** [comment]
@Baron-von-Riedesel was this fixed
also for you?

**Comment by Baron-von-Riedesel** [comment]
@Baron-von-Riedesel <https://github.com/Baron-von-Riedesel> was this 
fixed also for you?
>
> Message ID: 
> ***@***.***>
>

Can't remember - and currently I'm taking a long break from programming.
